langchain/callbacks/mlflow_callback.py

```diff
@@ -242,15 +243,22 @@ def __init__(
         """Initialize callback handler."""
         import_pandas()
         import_textstat()
-        import_mlflow()
+        mlflow = import_mlflow()
         spacy = import_spacy()
+
         super().__init__()
 
         self.name = name
         self.experiment = experiment
         self.tags = tags
         self.tracking_uri = tracking_uri
 
+        if "DATABRICKS_RUNTIME_VERSION" in os.environ:
+            with mlflow.start_run() as run:
+                experiment_id = run.info.experiment_id
+                experiment = mlflow.get_experiment(experiment_id)
+                self.experiment = experiment.name
+

```

Comment:
I suggest changing this block:

https://github.com/hwchase17/langchain/blob/c8d420231753a18657f094394175ccaad3b00baf/langchain/callbacks/mlflow_callback.py#L129-L148

to the following:
```python
    def __init__(self, **kwargs: Any):
        self.mlflow = import_mlflow()
        if "DATABRICKS_RUNTIME_VERSION" in os.environ:
            self.mlflow.set_tracking_uri("databricks")
        else:
	        tracking_uri = get_from_dict_or_env(
	            kwargs, "tracking_uri", "MLFLOW_TRACKING_URI", ""
	        )
	        self.mlflow.set_tracking_uri(tracking_uri)

	        # User can set other env variables described here
	        # > https://www.mlflow.org/docs/latest/tracking.html#logging-to-a-tracking-server

	        experiment_name = get_from_dict_or_env(
	            kwargs, "experiment_name", "MLFLOW_EXPERIMENT_NAME"
	        )
	        self.mlf_exp = self.mlflow.get_experiment_by_name(experiment_name)
	        if self.mlf_exp is not None:
	            self.mlf_expid = self.mlf_exp.experiment_id
	        else:
	            self.mlf_expid = self.mlflow.create_experiment(experiment_name)

        self.start_run(kwargs["run_name"], kwargs["run_tags"])
```
On databricks, experiments will be created when start_run is called in databricks python notebook, see https://docs.databricks.com/mlflow/experiments.html#create-notebook-experiment. So we don't need to bother handling experiment name and experiment creation here.

Reply:
Haven't tested these. Feel free to adjust if you see errors/typos. Thanks!

Reply:
`self.mlf_expid` is needed for downstream function calls, so we would still need to handle getting the experiment id

Reply:
Can we retrieve the experiment id by:
```python
run = mlflow.start_run()
self.mlf_expid = run.info.experiment_id
```

Reply:
Yes, but we need to have a `mlflow.end_run()` at the end...

Reply:
The code does this for logging:
```
    def metrics(
        self, data: Union[Dict[str, float], Dict[str, int]], step: Optional[int] = 0
    ) -> None:
        """To log all metrics in the input dict."""
        with self.mlflow.start_run(
            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
        ):
            self.mlflow.log_metrics(data)
```
Where every time they are logging something during every function call, the way they seem to go around this issue is by doing
```
self.run = self.mlflow.MlflowClient().create_run(
            self.mlf_expid, run_name=name, tags=tags
        )
```
instead of actually starting the run inside their start_run function...